### ISSUE TITLE
Extending material components

### ISSUE THREAD

**Issue by teterovic** [angular/material2#10208]
Use case: I want to disable a specific animation in the select component (the @transformPanel animation), in order to do so i just need to override the template or the animation property of the components annotation. #2194 #9216

Angular still doesn't support **Annotation inheritance**, but it's going to support it with the **Ivy Renderer / Angular v6 Compiler**.

Are there any contras to prevent this ? 

**CDK is nice but to construct a select component to be as good as the MatSelect component is a lot of work**, we used to copy the code from the original MatSelect component but **it's hard to maintain it with Material/CDK updates**.

Solving this problem would allow end-developer to reimplement individual functions of the base class/ component for example the new **disableOptionCentering** would not be needed because the end-developer could do it

Any best practice or **under development** for this case ?

**Comment by teterovic** [comment]
One solution to this problem is the following, so i can define my own animation or disable it in ./select-animations.ts or in the custom .html file. But it would be much much better if the @Component annotation could be inherited, the code would me much more clear.

`import { NgForm, NgControl } from '@angular/forms';
import { Directionality } from '@angular/cdk/bidi';
import { ViewportRuler } from '@angular/cdk/overlay';
import { ErrorStateMatcher, MAT_OPTION_PARENT_COMPONENT } from '@angular/material/core';
import { MatFormField, MatFormFieldControl } from '@angular/material/form-field';

import {
    NgZone,
    ChangeDetectorRef,
    Optional,
    Self,
    Attribute,
    Inject,
    ElementRef,
    Component,
    ViewEncapsulation,
    ChangeDetectionStrategy
} from '@angular/core';

import { FormGroupDirective } from '@angular/forms';
import { MAT_SELECT_SCROLL_STRATEGY, MatSelect } from '@angular/material';
import { matSelectAnimations } from './select-animations';

@Component({
    moduleId: module.id,
    selector: 'ab-select',
    exportAs: 'abSelect',
    templateUrl: 'select.component.html',
    inputs: ['disabled', 'disableRipple', 'tabIndex'],
    encapsulation: ViewEncapsulation.None,
    preserveWhitespaces: false,
    changeDetection: ChangeDetectionStrategy.OnPush,
    host: {
        'role': 'listbox',
        '[attr.id]': 'id',
        '[attr.tabindex]': 'tabIndex',
        '[attr.aria-label]': '_ariaLabel',
        '[attr.aria-labelledby]': 'ariaLabelledby',
        '[attr.aria-required]': 'required.toString()',
        '[attr.aria-disabled]': 'disabled.toString()',
        '[attr.aria-invalid]': 'errorState',
        '[attr.aria-owns]': 'panelOpen ? _optionIds : null',
        '[attr.aria-multiselectable]': 'multiple',
        '[attr.aria-describedby]': '_ariaDescribedby || null',
        '[attr.aria-activedescendant]': '_getAriaActiveDescendant()',
        '[class.mat-select-disabled]': 'disabled',
        '[class.mat-select-invalid]': 'errorState',
        '[class.mat-select-required]': 'required',
        'class': 'mat-select',
        '(keydown)': '_handleKeydown($event)',
        '(focus)': '_onFocus()',
        '(blur)': '_onBlur()',
    },
    animations: [
        matSelectAnimations.fadeInContent,
        matSelectAnimations.transformPanel
    ],
    providers: [
        { provide: MatFormFieldControl, useExisting: AbSelect },
        { provide: MAT_OPTION_PARENT_COMPONENT, useExisting: AbSelect }
    ],
})
// tslint:disable-next-line:component-class-suffix
export class AbSelect extends MatSelect {

    constructor(
        _viewportRuler: ViewportRuler,
        _changeDetectorRef: ChangeDetectorRef,
        _ngZone: NgZone,
        _defaultErrorStateMatcher: ErrorStateMatcher,
        elementRef: ElementRef,
        @Optional() _dir: Directionality,
        @Optional() _parentForm: NgForm,
        @Optional() _parentFormGroup: FormGroupDirective,
        @Optional() _parentFormField: MatFormField,
        @Self() @Optional() ngControl: NgControl,
        @Attribute('tabindex') tabIndex: string,
        @Inject(MAT_SELECT_SCROLL_STRATEGY) _scrollStrategyFactory) {
        // tslint:disable-next-line:max-line-length
        super(_viewportRuler, _changeDetectorRef, _ngZone, _defaultErrorStateMatcher, elementRef, _dir, _parentForm, _parentFormGroup, _parentFormField, ngControl, tabIndex, _scrollStrategyFactory);
    }
}
`

**Comment by NHKK** [comment]
Were you able to ended the components in the end without duplicate code that already existed in the library? I tried to "extend" the card component but it isn't as simple as my-mat-card extend mat-card.

**Comment by teterovic** [comment]
You have to look a lot at the sourcecode of the component. The bad thing is rlthat almost everyting is private and not protected.
Another solution is to fork the project on github and to maintain your version but i think this is more overhead than just to extend the component.

**Comment by andrewtw** [comment]
I believe this should be considered as a feature request.

**Comment by RoshanRoyal** [comment]
+1

**Comment by DenisSemionov** [comment]
+1

**Comment by dojoVader** [comment]
Yeah I spent days looking to extend a component, but as someone mentioned everything is private , it would be nice to be able to extend because a lot of the component don't fit our use scenario and not being able to extend is definitely a big problem for us.

**Comment by NHKK** [comment]
Could always fork and change everything to protected

**Comment by dojoVader** [comment]
> Could always fork and change everything to protected

I thought of that too but I don't know if we want to do that, I will have to consult with my team lead.
